packages/core/content-manager/admin/src/pages/EditView/components/InputRenderer.tsx

```diff
@@ -35,8 +45,54 @@ type InputRendererProps = DistributiveOmit<EditFieldLayout, 'size'>;
  * the complete EditFieldLayout and will handle RBAC conditions and rendering CM specific
  * components such as Blocks / Relations.
  */
-const InputRenderer = ({ visible, hint: providedHint, ...props }: InputRendererProps) => {
-  const { id, document, collectionType } = useDoc();
+const InputRendererWithContext = (props: InputRendererProps) => {
+  const id = useRelationModalContext('RelationModal', (state) => state.currentRelation.documentId);
+  const document = useRelationModalContext('RelationModal', (state) => state.document);
+  const model = useRelationModalContext('RelationModal', (state) => state.currentRelation.model);
+  const collectionType = useRelationModalContext(
+    'RelationModal',
+    (state) => state.currentRelation.collectionType
+  );
+  const changeCurrentRelation = useRelationModalContext(
+    'RelationModal',
+    (state) => state.changeCurrentRelation
+  );
+  const isModalOpen = useRelationModalContext('RelationModal', (state) => state.isModalOpen);
+  const {
+    edit: { components },
+  } = useDocumentLayout(model);
+
+  return (
+    <InputRenderer
+      contextId={id}
+      contextDocument={document}
+      contextCollectionType={collectionType}
+      contextModel={model}
+      changeCurrentRelation={changeCurrentRelation}
+      isModalOpen={isModalOpen}
+      contextComponents={components}
+      {...props}
+    />
+  );
+};
```

Comment:
This jumps out at me as strange. I feel like an issue with this current implementation is we have a context provider that is deep down the tree (in the modal). Wouldn't it be better if we had the relation context higher up the tree? Could we design something that provides the context for editing relations from the root entry down. That way no matter where we are we have access to whatever relation is being edited?

That's just an idea, not saying you should try to implement that right now. Happy to have a PP session with you and @remidej to come up with an API to support this feature.

Reply:
I agree that the Thing + ThingWithContext pattern is odd, I'd rather unify everything. Moving the context at the input level rather than the modal would be really nice indeed. And it's probably simpler to debug since there would be a single context rather than nested ones

Reply:
so your suggestion is to move everything in the Relations file?

Reply:
To give you some context, here is a representation of the flow of Components for the Relation field
`InputRenderer 
   -> UnstableRelationsInput aka UnstableRelationsField
	-> RelationsInput
	-> UnstableRelationsList
		-> UnstableListItem
			-> RelationModal
				-> RelationModalBody (CTX)
					-> FormLayout (usa il model)
						-> InputRendererWithContext
							-> InputRenderer
						-> InputRenderer`